src/python/pants/rules/core/binary.py

```diff
@@ -32,12 +35,26 @@ class CreatedBinary:
 
 
 @console_rule
-async def create_binary(addresses: BuildFileAddresses, console: Console, workspace: Workspace, options: Binary.Options) -> Binary:
+async def create_binary(
+    addresses: BuildFileAddresses,
+    console: Console,
+    workspace: Workspace,
+    options: Binary.Options,
+    options_bootstrapper: OptionsBootstrapper,
+    build_root: BuildRoot
+    ) -> Binary:
   with Binary.line_oriented(options, console) as print_stdout:
-    print_stdout("Generating binaries in `dist/`")
+    global_options = options_bootstrapper.bootstrap_options.for_global_scope()
+    pants_distdir = global_options.pants_distdir
+    relpath = os.path.relpath(pants_distdir, build_root.path)
+    if relpath.startswith("."):
+      print_stdout(f"pants-distdir `{pants_distdir}` must be set relative to the build root, cannot create binary.")
```

Comment:
I'm not sure this acts as intended:

```
In [1]: import os

In [2]: os.path.relpath("dist", "/usr/bin/pants")
Out[2]: '../../../Users/eric/DocsLocal/code/projects/pants/dist'

In [3]: os.path.relpath("/usr/bin/pants/dist", "/usr/bin/pants")
Out[3]: 'dist'
```

I think we want the user to be able to specify `--pants-distdir` as a relative path. Example 2 shows that this does not work. We also want to support absolute paths, which example 3 does show works.

Reply:
Instead, we could check 

```python
if Path(pants_distdir).is_absolute() and Path(pants_distdir).relative_to(buIld_root.path):
  print_stderr(f"When set to an absolute path, `--pants-distdir` must be relative to the build root. You set it to {}. Instead, use a relative path or an absolute path relative to the build root.")
```

Reply:
This won't cut it since a relative path can always escape the root via '..'. This should be a robust check:
```python
$ python3.6
Python 3.6.9 (default, Jul  8 2019, 20:56:32) 
[GCC 9.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pathlib import Path
>>> 
>>> 
>>> BUILD_ROOT = Path('/a/build/root')
>>> 
>>> 
>>> def is_legal(path):
...   abs_path = path if path.is_absolute() else BUILD_ROOT.joinpath(path).resolve()
...   return BUILD_ROOT == abs_path or BUILD_ROOT in abs_path.parents
... 
>>> 
>>> is_legal(BUILD_ROOT)
True
>>> is_legal(Path('/a/build/root/dist/dir'))
True
>>> is_legal(Path('dist/dir'))
True
>>> is_legal(Path('./dist/dir'))
True
>>> is_legal(Path('../root/dist/dir'))
True
>>> is_legal(Path('/other/root/dist/dir'))
False
>>> is_legal(Path('../not_root/dist/dir'))
False
>>> 
```

Reply:
I think that pathlib actually doesn't allow `../` when it computes `relative_to`:

```
>>> Path('/a/b').relative_to('/a')
PosixPath('b')
>>> Path('/a/b').relative_to('/c')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/benjyw/.pyenv/versions/3.7.4/lib/python3.7/pathlib.py", line 883, in relative_to
    .format(str(self), str(formatted)))
ValueError: '/a/b' does not start with '/c'
>>> ```